[1단계 - 장기] 코기(장재현) 미션 제출합니다.

jaehyeon2650 opened the pull request:
<!-- 

안녕하세요 아서! 이번 코드 리뷰를 부탁드리는 코기입니다 😊
구현하는데 시간이 오래걸려서 아쉬운 코드 상태로 PR를 제출합니다..🥹
잘 부탁드립니다!

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
[코기의 프롤로그](https://prolog.techcourse.co.kr/studylogs/4135) -> 아직 작성 못했습니다..

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
중복 코드가 많음

## 어떤 부분에 집중하여 리뷰해야 할까요?

- 시간이 부족하여 테스트에 신경을 쓰지 못했습니다. step1 피드백을 받고 테스트에 대한 리팩토링을 진행하겠습니다! 테스트 쪽은 지금 당장 피드백 안해주셔도 될거같습니다!
- Cannon, Chariot을 제외한 말들은 canMove의 메서드 구현이 똑같고 Cannon과 Chariot도 canMove 메서드 구현이 비슷합니다. 어떻게 중복 코드를 방지할 수 있을지 고민하다가 결국 해결 못 했습니다..
현재 코드에서는 `Movements`라는 클래스를 따로 두어 이동 로직의 중복을 제거하고 있습니다. 하지만 모든 추상 클래스 `Piece`의 자식 클래스들이 모든 `Movements`를 사용하는 것이 아닌, `차`와 `포` 객체는 이동로직이 달라서 `Movements`를 사용하고 있지 않습니다. 최대한 중복 코드를 제거하기 위해 `Movements`에 메서드를 추가하였는데, 지금과 같은 상황에서 기물들의 움직임에서 중복 코드를 제거하기 위해 억지로 분리한 느낌이 듭니다. 어떻게 중복 코드 문제를 해결할 수 있을까요? 🥹

junhaa committed:
docs: 요구사항 추가

junhaa committed:
feat: 위치 정보를 저장하는 객체 구현

- 보드판의 범위를 검증하는 기능 구현

jaehyeon2650 committed:
feat: solider 말 생성

- 말 이동 기능 추가

junhaa committed:
feat: 졸의 주변이 비어있을 때 이동 가능 여부를 판단하는 기능 구현

jaehyeon2650 committed:
feat: solider 이동 가능 여부 추가

jaehyeon2650 committed:
feat: 말 객체 생성

junhaa committed:
feat: 사용자의 입력에 맞게 상, 마의 배치를 변경하는 기능 구현

jaehyeon2650 committed:
feat: rightElephant 상 마 위치 구현

jaehyeon2650 committed:
feat: LeftElephant 상 마 위치 구현

jaehyeon2650 committed:
feat: 말 위치 초기화 기능 추가

junhaa committed:
feat: 장기판 초기화 기능 구현

junhaa committed:
feat: 최종 목적지의 좌표를 반환하는 기능 구현

jaehyeon2650 committed:
feat: 중간 이동 지점 반환 기능 추가

junhaa committed:
feat: 최종 목적지로 이동이 가능한지 판단하는 기능 구현

junhaa committed:
feat: 이동 경로 지점을 반환하는 기능 구현

jaehyeon2650 committed:
feat: 졸이 이동할 수 있는지 판단하는 로직 추가

junhaa committed:
feat: 마의 이동 가능 여부를 반환하는 기능 구현

jaehyeon2650 committed:
feat: 코끼리 말이 이동할 수 있는지 판단하는 로직 추가

junhaa committed:
feat: 장군의 이동 가능 여부를 반환하는 기능 구현

junhaa committed:
feat: 선비의 이동 가능 여부를 반환하는 기능 구현

jaehyeon2650 committed:
feat: 전차 말이 이동할 수 있는지 판단하는 로직 추가

junhaa committed:
feat: 포가 이동할 수 있는지 판단하는 로직 추가

jaehyeon2650 committed:
feat: 말을 움직이는 기능 추가

junhaa committed:
feat: 게임 승리 팀을 반환하는 기능 구현

jaehyeon2650 committed:
feat: 승자를 조회하는 기능 추가

junhaa committed:
feat: 게임 초기화 과정 구현

jaehyeon2650 committed:
feat: 장기말 출력 기능 추가

junhaa committed:
feat: 장기 게임 실행 로직 구현

junhaa committed:
fix: 포 이동 오류 해결

jaehyeon2650 committed:
feat: 이름 검증 추가

jaehyeon2650 committed:
refactor: 장기 말 상태 예외 발생 순서 변경

junhaa committed:
feat: 사용자의 좌표 입력 검증 추가

jaehyeon2650 committed:
Merge remote-tracking branch 'origin/step1' into step1

junhaa committed:
refactor: view 내부 메서드 추출

jaehyeon2650 committed:
refactor: JanggiRunner 내부 메서드 추출

junhaa committed:
refactor: 코드 내부 메서드 추출

jaehyeon2650 committed:
feat: 같은 종류의 말인지 확인하는 기능 추가

jaehyeon2650 committed:
refactor: Board 메서드 추출

jaehyeon2650 committed:
refactor: 상 말 패턴 추상클래스로 변경

jaehyeon2650 committed:
refactor: PieceFactory 메서드 분리

jaehyeon2650 committed:
refactor: 움직임 판단 메서드 이동

jaehyeon2650 committed:
refactor: 도착지점에 아군인지 판단하는 메서드 이동

jaehyeon2650 committed:
refactor: Cannon 메서드 추출

junhaa committed:
refactor: 차, 포 내부 메서드 분리

Hyunta commented on src/main/java/domain/JanggiRunner.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.List;
+import view.InputView;
+import view.OutputView;
+
+public class JanggiRunner {
```
JanggiRunner와 JanggiGame을 두개로 나누신 이유가 궁금합니다.

jaehyeon2650 replied:
JanggiGame은 Board와 Player 상태를 가져 전반적을 게임에 필요한 필수적인 기능을 수행하도록 하였습니다. JanggiRunner는 InputView, OutputView와 JanngiGame을 연결하여 게임의 전체적인 흐름(초, 한나라 순서로 게임을 진행 등)을 담당하도록 하였습니다.
이를 분리하지 않으면 의존성도 많아지고 게임 상태 관리, 게임 흐름을 담당 등 많은 역할이 객체에 부여될거같아 분리하게 되었습니다!

Hyunta commented on src/main/java/domain/JanggiRunner.java:
```diff
@@ -0,0 +1,93 @@
+        return SetupOption.findSetupStrategy(firstPlayerOption);
+    }
+
+    private Players createPlayers() {
+        Usernames usernames = createUsernames();
+        String startPlayerName = inputView.getStartPlayerName();
+        return Players.createFrom(usernames, startPlayerName);
+    }
+
+    private Usernames createUsernames() {
+        String firstPlayerName = inputView.getFirstPlayerName();
+        String secondPlayerName = inputView.getSecondPlayerName();
+        return new Usernames(firstPlayerName, secondPlayerName);
+    }
```
UserNames, Players를 생성해주는 로직이면 각 객체에서 관리해주면 되지 않을까요?

jaehyeon2650 replied:
정확히 어떤 관리 방식인지 이해가 되지 않네요..🥹 혹시 다음과 같은 방법일까요??
```
public class Usernames {
    private final String playerAName;
    private final String playerBName;

    public Usernames(String playerAName, String playerBName) {
        validateDuplicate(playerAName, playerBName);
        this.playerAName = playerAName;
        this.playerBName = playerBName;
    }

     public static Usernames from(InputView inputView) {
        return new Usernames(inputView.getFirstPlayerName(), inputView.getSecondPlayerName());
    }
}
```
만약 위와 같이 바꾸게 된다면 InputView 구현체를 파라미터로 받지 않고 InputView 관련 인터페이스를 만들어서 view가 변경이 되어도 대응하기 쉽게 만들거같습니다!

Hyunta replied:
기존에는 JangiRunner가 각 객체를 생성해주고 있습니다.
제가 말해드린 방법은 각 객체가 주어진 파라미터를 받아서 스스로 생성하는 방식입니다.
InputView를 넘겨서는 안되겠죠, InputView에서 입력받은 값만 넘겨서 생성하도록 하는것이 적절해보입니다.

Hyunta commented on src/main/java/domain/JanggiRunner.java:
```diff
@@ -0,0 +1,93 @@
+    private HorseElephantSetupStrategy chooseStrategy(String players) {
+        String firstPlayerOption = inputView.getSetupNumber(players);
+        return SetupOption.findSetupStrategy(firstPlayerOption);
+    }
```
```suggestion
    private HorseElephantSetupStrategy chooseStrategy(String playerName) {
        String firstPlayerOption = inputView.getSetupNumber(playerName);
        return SetupOption.findSetupStrategy(firstPlayerOption);
    }
```

jaehyeon2650 replied:
파라미터 이름에 신경쓰지 못한 부분이 많군요..! 이 외에도 이해가 되지 않는 부분은 수정하도록 하겠습니다! 지적 감사합니다 😊

Hyunta commented on src/main/java/domain/JanggiRunner.java:
```diff
@@ -0,0 +1,93 @@
+    private JanggiGame initializeGame() {
+        Players players = createPlayers();
+        HorseElephantSetupStrategy firstPlayerStrategy = chooseStrategy(players.getChoPlayerName());
+        HorseElephantSetupStrategy secondPlayerStrategy = chooseStrategy(players.getHanPlayerName());
```
cho, han으로 할지 first, second로 할지 통일해주세요

jaehyeon2650 replied:
다른 개발자 분이 보시면 헷갈릴거 같습니다! 통일 하도록 하겠습니다 👍

Hyunta commented on src/main/java/domain/JanggiRunner.java:
```diff
@@ -0,0 +1,93 @@
+
+    private List<Piece> createAllPieces(HorseElephantSetupStrategy firstPlayerStrategy,
+                                        HorseElephantSetupStrategy secondPlayerStrategy) {
+        PieceFactory factory = new PieceFactory();
+        return factory.createAllPieces(firstPlayerStrategy, secondPlayerStrategy);
```
PieceFactory를 인스턴스로 관리하신 이유가 있을까요? static 메서드로 호출해주면 안되나요?

jaehyeon2650 replied:
저도 이 부분을 고민하였습니다. 상태가 없는 객체들은 무조건 static 클래스로 만들어야하는가? 상태가 없는 객체는 있으면 안되는가? 고민하였습니다.
아직 정확한 저만의 기준은 못 찾았습니다. 현재 정한 기준은 여러 곳에서 사용하는가? 입니다. 보통 static 클래스는 여러 객체에서 사용할 수 있을거 같은 경우 util 패키지에 관리합니다. 하지만 현재 PieceFactory는 말의 위치를 초기화하는 기능을 하는데 이 기능이 장기 게임 초기화 말고는 다른 상황에 쓰이지 않을거같아서 static 메서드로 만들지 않았습니다.
이 기준이 좋은 기준인지는 아직 모르겠습니다..🥹 아서는 어떤 기준으로 static 메서드를 만드나요? 또한 상태가 없는 객체에 대해 어떻게 생각하시나요?

Hyunta replied:
맞아요, 어떤 경우에 static 으로 관리하는 것이 적절한지 모호합니다.
객체를 인스턴스로 관리하는 것과 static 메서드를 호출하는 것에는 어떤 차이가 있나요?
결론부터 말씀드리면 static으로 관리하는 것과 인스턴스를 호출해서 사용하는 것은 큰 차이는 없습니다.
다만, 각각 JVM에서 어떻게 관리되어지는지 정도만 알고가면 좋을 것 같습니다.

Hyunta commented on src/main/java/domain/piece/PieceFactory.java:
```diff
@@ -0,0 +1,56 @@
+        return Stream.of(choElephantHorse, hanElephantHorse, hanDefaultPieces, choDefaultPieces)
+                .flatMap(List::stream)
+                .toList();
+    }
```
초기 기물을 생성하는 부분을 클래스로 분리하신 이유가 있을까요?

jaehyeon2650 replied:
JanngiGame이나 Board에서 초기화하는 방법이 가장 처음에 생각났습니다! 하지만 초기화하는 과정에서 말의 위치의 하드코딩이 상당히 많아 클래스 내부가 지저분해진다는 느낌을 받았습니다. 그래서 초기 기물을 생성하는 부분을 클래스로 따로 나눴습니다. 
또한 JanngiGame이나 Board 생성자에서 내부 초기화를 하면 테스트하기도 어려울거 같아 List<Piece>로 생성자를 구성하고 원하는 위치 상태에서 Piece들을 테스트할 수 있게 구성하였습니다.

Hyunta commented on src/main/java/domain/Board.java:
```diff
@@ -0,0 +1,83 @@
+
+    public List<Piece> getAlivePieces() {
+        return alivePieces.stream().map(Piece::newInstance).toList();
+    }
```
getter에서 Piece를 새로 생성해서 반환하는 이유가 있을까요?

jaehyeon2650 replied:
piece에는 moveTo 메서드가 존재합니다. moveTo 메서드는 Piece 객체 상태를 변경합니다. 그렇기에 만약 새로 생성하지 않고 반환을 한다면 다른 개발자가 getter를 사용하고 실수로 moveTo를 호출하여 게임 안에 있는 말들에 영향을 미칠거라고 생각하여 복사하여 반환하였습니다!

Hyunta commented on src/main/java/domain/Board.java:
```diff
@@ -0,0 +1,83 @@
+
+    public List<Piece> getAlivePieces() {
+        return alivePieces.stream().map(Piece::newInstance).toList();
+    }
```
다른 부분과 컨벤션을 맞춰주세요.

```suggestion
    public List<Piece> getAlivePieces() {
        return alivePieces.stream()
        .map(Piece::newInstance)
        .toList();
    }
```

jaehyeon2650 replied:
앗 놓친 부분이네요! 다른 부분과 컨벤션을 맞추도록 하겠습니다! 지적 감사합니다😊

Hyunta commented on src/main/java/domain/piece/Piece.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.List;
+import java.util.function.Predicate;
+
+public abstract class Piece {
+    protected Position position;
+    protected final TeamType teamType;
```
Piece의 Position을 Board에서 관리해줄 수도 있었을 것 같은데, Piece에서 관리하기로 결정하신 이유가 궁금합니다.

jaehyeon2650 replied:
처음에는 저도 아서와 같은 의견으로 Board에서 Position을 Map<Position, Piece> 이렇게 관리하도록 하려했습니다. 
이럴 경우 보드가 각 기물마다 움직이는 방법을 알아야하고 각 Piece가 어떤 구현체인지 알기 위해서는 getType 혹은 instanceOf를 사용하여 어떤 기물인지 알아야합니다. 즉 Board에서 Piece라는 추상 클래스만 아는 것이 아니라 구체 클래스까지 알고 각 구체 클래스마다 어떻게 이동이 되는지 아는 것이 부자연스럽다고 느껴집니다.
또한 그럴 일은 거의 없겠지만 특정 기물의 이동 경로가 변경이 된다면 각 기물이 어떻게 이동하는지 Board가 알고 있기에 Board를 수정할 것입니다. 특정 기물의 이동 경로가 변경되면 해당 기물 클래스에서 수정이 이뤄질거같은데 Board를 수정하는게 부자연스럽다고 느꼈습니다

Hyunta replied:
Board에서 Positon을 관리하면서 기물의 움직이는 방법을 Piece에서 관리할 수 있을 것 같은데요.
위치만 Board가 가지고 있고 이동하는 로직은 Piece에게 물어보면 되지 않나요?

Hyunta commented on src/main/java/domain/Position.java:
```diff
@@ -0,0 +1,42 @@
+
+    public boolean canMovePosition(int deltaRow, int deltaColumn) {
+        return row.canMoveRow(deltaRow) && column.canMoveColumn(deltaColumn);
+    }
```
canMovePosition은 MovePostion을 할 때 검증하도록 구현하면 안되나요?

jaehyeon2650 replied:
movePosition에서 row.moveRow, column.moveColumn 내부에서 검증을 진행합니다.
그럼에도 따로 검증하는 canMovePostion 메서드를 뺀 이유는 외부에서 예외 처리를 하는 상황을 최소화하기 위해서 입니다. movePoistion 하기전에 움직일 수 있는지 체크하여 가능한 경우에만 movePosition을 하도록 하여 movePosition에서 발생하는 예외를 사전에 방지하는 식으로 구현하였습니다!

Hyunta replied:
movePosition을 이용해서 개발하는 동료들도 canMovePosition이 선행조건이라는 것을 알아야 합니다.
꼭 검증이 필요한 부분이라면 movePosition 내부에서 검증 로직을 호출하는 방식으로 하면 더 안전하지 않나요?

Hyunta commented on src/main/java/domain/Usernames.java:
```diff
@@ -0,0 +1,33 @@
+package domain;
+
+public class Usernames {
+    private final String playerAName;
+    private final String playerBName;
```
사용자의 이름을 왜 이렇게 따로따로 관리하시게 되셨나요?

jaehyeon2650 replied:
> 첫번째 플레이어의 이름을 입력해주세요.
> 코기
> 
> 두번째 플레이어의 이름을 입력해주세요.
> 루키
> 
> 먼저 시작할 플레이어의 이름을 입력해주세요.
> 루키

현재 위와 같이 입력을 받고 있습니다. 플레이어 이름이 중복되는 것을 검증하고 싶었습니다. 하지만 Player 객체에는 username과 teamType 필드를 가지고 있고 이를 생성하기 위해서는 세번째 입력까지 모두 받아야가능했습니다.(username과 teamType이 final 필드이기에 초기화때 무조건 필요합니다. )
따로 관리를 안하면 사용자가 중복된 플레이어의 이름을 두 번 입력해도 예외가 발생하지 않고 먼저 시작할 플레이어의 이름을 입력할 때 검증을 진행하여 늦게 예외 메시지를 받게 됩니다.
이런 예외 메시지가 두번째 플레이어 입력 때 바로 나오는게 흐름상 맞다고 판단하였습니다. 그래서 Usernames라는 객체를 만들어 두번째 입력이 끝나면 해당 객체를 통해 예외가 발생하도록 설계하였습니다.
이 예외를 inputview에서 담당하는 것은 비즈니스 로직 예외를 뷰에서하는 건 아니라 판단하여 따로 예외 처리하기 위해 만들었습니다.

Hyunta commented on src/main/java/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    @Override
+    public boolean canMove(Position expectedPosition, List<Piece> pieces) {
+        Direction direction = findDirectionToReachAt(expectedPosition);
+        if (direction == null) {
+            return false;
+        }
+        if (hasCannonPieceAtIntermediatePositions(expectedPosition, pieces, direction)) {
+            return false;
+        }
+        if (hasOnlyOnePieceAtIntermediatePositions(expectedPosition, pieces, direction)) {
+            return false;
+        }
+        return hasNotTeamAtPosition(expectedPosition, pieces, (piece -> piece.isSameType(PieceType.CANNON)));
+    }
```
> Cannon, Chariot을 제외한 말들은 canMove의 메서드 구현이 똑같고 Cannon과 Chariot도 canMove 메서드 구현이 비슷합니다. 어떻게 중복 코드를 방지할 수 있을지 고민하다가 결국 해결 못 했습니다..
현재 코드에서는 Movements라는 클래스를 따로 두어 이동 로직의 중복을 제거하고 있습니다. 하지만 모든 추상 클래스 Piece의 자식 클래스들이 모든 Movements를 사용하는 것이 아닌, 차와 포 객체는 이동로직이 달라서 Movements를 사용하고 있지 않습니다. 최대한 중복 코드를 제거하기 위해 Movements에 메서드를 추가하였는데, 지금과 같은 상황에서 기물들의 움직임에서 중복 코드를 제거하기 위해 억지로 분리한 느낌이 듭니다. 어떻게 중복 코드 문제를 해결할 수 있을까요? 🥹

지금은 Movement에서도 움직임에 대한 검증을 하고 있고, 각 기물에서도 움직임에 대한 검증을 하고 있는 것 같은데요.
Movenment라는 객체가 딱히 왜 필요한지 제가 잘 모르겠어서 설명을 해주실 수 있을까요?
Board에서 기물들의 위치를 관리하는 방향으로 구현한다면 중복을 조금 개선해볼 수 있지 않을까요?

Hyunta reviewed (changes requested):
안녕하세요 코기,
장기 미션으로 만나게된 아서라고 합니다.
몇가지 커멘트 남겨드렸으니 확인 부탁드립니다.

jaehyeon2650 reviewed (commented):
시간내서 리뷰를 해주셔서 감사합니다 😊
코멘트에 대해 이해하지 못한 부분은 개인적으로 dm 드렸습니다! dm 내용을 바탕으로 다시 답변하도록 하겠습니다!

jaehyeon2650 committed:
refactor: 메서드 파라미터 이름 변경

jaehyeon2650 committed:
feat: PathFinder 인터페이스 생성 및 구현체 생성

- 목적지까지 가는 경로 반환하는 기능
- 가는 방법이 고정되어 있는 경우 구현체
- 가는 방법이 방향만 있는 경우 구현체

jaehyeon2650 committed:
feat: PathValidator 인터페이스 생성 및 구현체 생성

- 특정 피스가 해당 목적지까지 갈 수 있는지 검증하는 로직 추가

jaehyeon2650 committed:
refactor: Position 위치 이동

- board에서 Position 관리
- board는 살아있는 말과 위치를 관리
- board는 말에게 메시지를 보내 이동 검증을 진행
- Position 이동에 따른 outputView 수정

jaehyeon2650 committed:
test: position fixture 생성 및 적용

jaehyeon2650 committed:
test: 이동 관련 검증 테스트

jaehyeon2650 committed:
refactor: 패키지 이동

jaehyeon2650 committed:
Merge pull request #1 from jaehyeon2650/step1-1

리팩토링

Hyunta reviewed (approved):
안녕하세요 코기,
피드백 반영해주신 부분 확인했습니다.
남겨주신 질문에 답변을 달아드렸는데 다음 단계 진행하시면서 이어서 고민해주세요.
